src/iop/retouch.c

```diff
@@ -0,0 +1,4569 @@
+}
+
+static gboolean rt_wdbar_button_release(GtkWidget *widget, GdkEventButton *event, dt_iop_module_t *self)
+{
+  dt_iop_retouch_gui_data_t *g = (dt_iop_retouch_gui_data_t *)self->gui_data;
+  if(event->button == 1)
+  {
+    g->is_dragging = 0;
+  }
+  return TRUE;
+}
+
+static gboolean rt_wdbar_scrolled(GtkWidget *widget, GdkEventScroll *event, dt_iop_module_t *self)
+{
+  if(darktable.gui->reset) return TRUE;
+  
+  dt_iop_request_focus(self);
+  
+  int delta_y;
+  if(dt_gui_get_scroll_unit_deltas(event, NULL, &delta_y))
+  {
+    dt_iop_retouch_params_t *p = (dt_iop_retouch_params_t *)self->params;
+    dt_iop_retouch_gui_data_t *g = (dt_iop_retouch_gui_data_t *)self->gui_data;
+    
+    const int inset = RT_WDBAR_INSET;
+
+    GtkAllocation allocation;
+    gtk_widget_get_allocation(widget, &allocation);
+    const float height = allocation.height - 2 * inset;
+
+    gboolean is_under_mouse = 0;
+    
+    if (!is_under_mouse)
+    {
+      is_under_mouse = rt_mouse_over_bottom_wdbar(g->mouse_y, height);
+      if (is_under_mouse)
+      {
+        const int num_scales = (p->num_scales - delta_y);
+        rt_num_scales_update(num_scales, self);
+      }
+    }
+    
+    if (!is_under_mouse)
+    {
+      is_under_mouse = rt_mouse_over_top_wdbar(g->mouse_y);
+      if (is_under_mouse)
+      {
+        const int merge_from_scale = (p->merge_from_scale - delta_y);
+        rt_merge_from_scale_update(merge_from_scale, self);
+      }
+    }
+    
+    if (!is_under_mouse)
+    {
+      is_under_mouse = 1;
+      const int curr_scale = (p->curr_scale - delta_y);
+      rt_curr_scale_update(curr_scale, self);
+    }
+  }
+
+  return TRUE;
+}
+
+static gboolean rt_wdbar_motion_notify(GtkWidget *widget, GdkEventMotion *event, dt_iop_module_t *self)
+{
+  dt_iop_retouch_gui_data_t *g = (dt_iop_retouch_gui_data_t *)self->gui_data;
+  
+  const int inset = RT_WDBAR_INSET;
+  
+  GtkAllocation allocation;
+  gtk_widget_get_allocation(widget, &allocation);
+  const float width = allocation.width - 2 * inset;
+  const float height = allocation.height - 2 * inset;
+
+  /* record mouse position within control */
+  g->mouse_x = CLAMP(event->x - inset, 0, width);
+  g->mouse_y = CLAMP(event->y - inset, 0, height);
+
+  if (g->is_dragging == dt_iop_retouch_drag_bottom)
+  {
+    const int num_scales = rt_mouse_x_to_wdbar_box(g->mouse_x, width);
+    rt_num_scales_update(num_scales, self);
+  }
+  
+  if (g->is_dragging == dt_iop_retouch_drag_top)
+  {
+    const int merge_from_scale = rt_mouse_x_to_wdbar_box(g->mouse_x, width);
+    rt_merge_from_scale_update(merge_from_scale, self);
+  }
+  
+  gtk_widget_queue_draw(self->widget);
+
+  return TRUE;
+}
+
+static int rt_scale_has_shapes(dt_iop_retouch_params_t *p, const int scale)
+{
+  int has_shapes = 0;
+  
+  for (int i = 0; i < RETOUCH_NO_FORMS && has_shapes == 0; i++)
+  {
+    has_shapes = (p->rt_forms[i].formid != 0 && p->rt_forms[i].scale == scale);
+  }
+
+  return has_shapes;
+}
+
+#if 0
```

Comment:
I would remove the code or add a comment about why it will be used later. This is large code and it will certainly ease the integration if retouch.c get smaller.